### PR TITLE
[Remove] Type from nested fields using new metadata field mapper

### PR DESCRIPTION
**Pull request by nknize**
types support is removed yet nested documents use the `_type` field to store the
path for nested documents. A new `_nested_path` metadata field mapper is added to
take the place of the `_type` field in order to remove the type dependency in
nested documents. BWC is handled in the new field mapper to ensure compatibility
with older versions.

relates #1940 
relates #2989

### CODE REVIEW

File: server/src/main/java/org/opensearch/indices/mapper/MapperRegistry.java
Comment (reta) [inline server/src/main/java/org/opensearch/indices/mapper/MapperRegistry.java:84]: Shouldn't it be `Version.V_2_0_0` ?

File: server/src/main/java/org/opensearch/index/mapper/NestedPathFieldMapper.java
Comment (reta) [inline server/src/main/java/org/opensearch/index/mapper/NestedPathFieldMapper.java:27]: Just an opinion, `_nested_path` looks too verbose, may be just `_path`?

File: server/src/main/java/org/opensearch/index/mapper/ObjectMapper.java
Comment (reta) [inline server/src/main/java/org/opensearch/index/mapper/ObjectMapper.java:420]: This is a bit risky call: when fixing one of the `Joda` conversions, realized that `settings` may not hold the index created version all the time [1], may be worth checking here as well.

[1] https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/mapper/Mapper.java#L258

File: server/src/main/java/org/opensearch/indices/mapper/MapperRegistry.java
Comment (nknize) [reply server/src/main/java/org/opensearch/indices/mapper/MapperRegistry.java:84]: Yes. Good catch

File: server/src/main/java/org/opensearch/index/mapper/MapperService.java
Comment (dreamer-89) [inline server/src/main/java/org/opensearch/index/mapper/MapperService.java:165]: As `NestedPathFieldMapper.NAME` is introduced in `2.0`; do we need this as meta field in < 7.8 version ?

**Comment by CEHENKLE** [comment]
Hey @nknize Just double checking -- I do not think you meant for this to get pulled into 2.0.0-rc1.  If anything, we would fix it between 2.0.0-rc1 and 2.0.0, right?

**Comment by nknize** [comment]
> between 2.0.0-rc1 and 2.0.0, right?

Correct. We just need to get it in for 2.0.0 along with TypeFieldMapper removal.

File: server/src/main/java/org/opensearch/index/mapper/NestedPathFieldMapper.java
Comment (nknize) [reply server/src/main/java/org/opensearch/index/mapper/NestedPathFieldMapper.java:27]: I ended up sticking with `_nested_path` because it's more descriptive when analyzing lucene segments (which has no object->field relationship). The fieldType is easily identifiable as the path for a nested field when just reading the field name.

File: server/src/main/java/org/opensearch/index/mapper/ObjectMapper.java
Comment (nknize) [reply server/src/main/java/org/opensearch/index/mapper/ObjectMapper.java:420]: Did you find this to be true in production or could it have just been a test that didn't set the Version? I don't think we can have a case where an index is created without a version, otherwise rolling upgrades would fail.

File: server/src/main/java/org/opensearch/index/mapper/MapperService.java
Comment (nknize) [reply server/src/main/java/org/opensearch/index/mapper/MapperService.java:165]: Alas, no! 2.0 is only bwc to 7.10.0 indexes. Thx!

File: server/src/main/java/org/opensearch/index/mapper/ObjectMapper.java
Comment (reta) [reply server/src/main/java/org/opensearch/index/mapper/ObjectMapper.java:420]: It is this one https://github.com/opensearch-project/OpenSearch/pull/2094/files

File: server/src/main/java/org/opensearch/index/mapper/NestedPathFieldMapper.java
Comment (reta) [reply server/src/main/java/org/opensearch/index/mapper/NestedPathFieldMapper.java:27]: :+1:

File: server/src/main/java/org/opensearch/index/mapper/ObjectMapper.java
Comment (nknize) [reply server/src/main/java/org/opensearch/index/mapper/ObjectMapper.java:420]: Just getting back to this. The [testDefaultConfig](https://github.com/opensearch-project/OpenSearch/pull/3004/files#diff-7b7553a5370e5b27cb2427f71bd8a727467b91fe30fca0ef321e440c0596f577R25) I had in there would've failed since it uses prefabricated empty settings but it was updating w/ different mappings so I added an [explicit test](https://github.com/opensearch-project/OpenSearch/pull/3004/files#diff-7b7553a5370e5b27cb2427f71bd8a727467b91fe30fca0ef321e440c0596f577R37) based on #1921 to update with the same mappings and Empty settings. There are no issues; we may revisit if this rears it's head again.

**Review by dreamer-89 — APPROVED** [review]
Thank you @nknize for picking up this tricky change. LGTM!

**Comment by nknize** [comment]
I'm going to give this until COB today and will merge if there are no objections. We need to get this in.

**Review by dblock — APPROVED** [review]

**Pull request merged**
